### PR TITLE
[enocean] Add null annotations

### PR DESCRIPTION
**Pull request by lsiepel**
A larger PR, hope you don't get overwhelmed with the amount of changed files, it is due to the nature of Enocean equipment profiles as part of the protocol. Most of those classes are annotated and did not need any further changes.
This PR mainly focuses on fixes most SAT-warnings. I think it is better to leave the eep package names as it is a clear structure. If that needs to be revised too, please suggest how and I’ll do the work.
-	Null annotations
-	checkstyle fixes
-	some improved logging / handling nothing functional for end-user.

When reviewing, the main files to look after are the non- EEP files, especially the handlers and discovery class. I tried to be careful and not change actual workings of the binding, but I might have missed something, so before merging I think it is good to have some additional testing. The JAR can be found here: 
https://1drv.ms/u/s!AnMcxmvEeupwjplBysVJv_nBA-mtfg?e=591tvj (version 4.0.0  2022-12-26 22.53)

Some issues I find hard to solve, are about compiler warnings and documentation being off. (or just my limited brains).  One example:

```
   protected final Hashtable<RORG, EEPType> receivingEEPTypes = new Hashtable<>();

   @Nullable
   EEPType existingKey = receivingEEPTypes.putIfAbsent(receivingEEPType.getRORG(), receivingEEPType);
   if (existingKey != null) {
      throw new IllegalArgumentException("Receiving more than one EEP of the same RORG is not supported");
   }
```
Warning: Redundant null check: The variable existingKey cannot be null at this location.
I wonder why, because the `putIfAbsent` returns null if the key is not in the receivingEEPTypes. I have some more, so i would appreciate some guidance.

### CODE REVIEW

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/discovery/EnOceanDeviceDiscoveryService.java [hunk 1]
```diff
@@ -183,8 +179,9 @@ public void eventReceived(EventMessage event) {
         }
     }
 
-    private Integer sendTeachInResponse(ERP1Message msg, String enoceanId) {
+    private @Nullable Integer sendTeachInResponse(ERP1Message msg, String enoceanId) {
         // get new sender Id
+        @Nullable
```
Comment (fwolter) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/discovery/EnOceanDeviceDiscoveryService.java:184]: The compiler is intelligent enough on local variables, so this could be removed.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/discovery/EnOceanDeviceDiscoveryService.java [hunk 2]
```diff
@@ -193,9 +190,12 @@ private Integer sendTeachInResponse(ERP1Message msg, String enoceanId) {
             // send response
             EEP response = EEPFactory.buildResponseEEPFromTeachInERP1(msg, newSenderId, true);
             if (response != null) {
-                bridgeHandler.sendMessage(response.getERP1Message(), null);
-                logger.debug("Teach in response for {} with new senderId {} (= offset {}) sent", enoceanId,
-                        HexUtils.bytesToHex(newSenderId), offset);
+                BasePacket bPacket = response.getERP1Message();
+                if (bPacket != null) {
+                    bridgeHandler.sendMessage(bPacket, null);
```
Comment (fwolter) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/discovery/EnOceanDeviceDiscoveryService.java:195]: Unfortunately the original code is using null values for signalling exceptions, which makes this new code more verbose. Did you thought about fixing this all at once? Otherwise, there will be another big PR refactoring the code to make use of exceptions.

**Review by fwolter — CHANGES_REQUESTED** [review]

**Comment by fwolter** [comment]
> Warning: Redundant null check: The variable existingKey cannot be null at this location.
> I wonder why, because the `putIfAbsent` returns null if the key is not in the receivingEEPTypes.

I agree that the return value of `putIfAbsent` can be null. @J-N-K is active at lastNPE. Maybe he has an idea, if this is a bug?

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/discovery/EnOceanDeviceDiscoveryService.java (hunk 2 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/discovery/EnOceanDeviceDiscoveryService.java:194]: There is certainly room for improvement, but this is beyond the scope of this PR. I might look into the exception handling in the future, but for that i would need more Enocean devices to test them myself.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanHandlerFactory.java [hunk 3]
```diff
@@ -56,12 +63,15 @@ public class EnOceanHandlerFactory extends BaseThingHandlerFactory {
     private Map<ThingUID, ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
 
     @Reference
+    @Nullable
     SerialPortManager serialPortManager;
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanHandlerFactory.java]: I have changed it, could you look at it again? I'm not very familiair with the OSGI framework.

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/discovery/EnOceanDeviceDiscoveryService.java (hunk 2 above)
Comment (fwolter) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/discovery/EnOceanDeviceDiscoveryService.java:194]: I started to refactor another binding to add null annotations myself in the past ... and aborted. I realized that one needs to touch the whole code, if the binding wasn't written with null annotations in mind. Otherwise you end up with adding null checks where they aren't neccessary at all by design, but demanded by the compiler (injection, exceptions or interfaces declaring nullable, although [only one call has a nullable argument.](https://github.com/openhab/openhab-addons/pull/14023/files#diff-557e3fbed7d8ee77148027c1cac66f57de95ef5d3344490f2f6f86f734d0c180R275) Here: pushing down the channel configuration the entire call hierarchy and checking each occurrence in every implementing class for null). This bloats up the code, is prone to introduce regressions and is hard to review, just to fix the compiler warnings. Especially corner cases where something *can* be null are hard to test, though.

My approach is to leave code without null annotations as-is, unless you add features and do a big refactoring anyway. 

Just for clarity, I think null annotations are a good thing and don't bloat up the code, but only if the code is designed with null annotations in mind.

My suggestion for proceeding with this PR is to make further improvements in a follow-up PR, if you want to and only address the review feedback in this PR.

However, fixing the other checkstyle warnings (naming schemes) is a good thing!

**Comment by fwolter** [comment]
The folder of the generic package is still upper case.

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/discovery/EnOceanDeviceDiscoveryService.java (hunk 2 above)
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/discovery/EnOceanDeviceDiscoveryService.java:194]: Thanks, i keep that in mind with other bindings that are on my list of adding null annotations. For this binding i'll look to get some devices and do more refactoring in another PR.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanHandlerFactory.java [hunk 4]
```diff
@@ -36,66 +38,76 @@
...
     public boolean supportsThingType(ThingTypeUID thingTypeUID) {
         return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
     }
 
     @Override
-    protected ThingHandler createHandler(Thing thing) {
+    protected @Nullable ThingHandler createHandler(Thing thing) {
         ThingTypeUID thingTypeUID = thing.getThingTypeUID();
-
+        ItemChannelLinkRegistry localItemChannelLinkRegistry = itemChannelLinkRegistry;
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanHandlerFactory.java:88]: Become obsolete after adressing other feedback.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanHandlerFactory.java [hunk 5]
```diff
@@ -36,66 +38,76 @@
...
 
     @Override
     protected void removeHandler(ThingHandler thingHandler) {
-        if (this.discoveryServiceRegs != null) {
-            ServiceRegistration<?> serviceReg = this.discoveryServiceRegs.get(thingHandler.getThing().getUID());
-            if (serviceReg != null) {
-                serviceReg.unregister();
-                discoveryServiceRegs.remove(thingHandler.getThing().getUID());
-            }
+        @Nullable
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanHandlerFactory.java:106]: Added this to prevent the info message:
`Unsafe interpretation of method return type as '@NonNull' based on the receiver type '@NonNull Map<@NonNull ThingUID,@NonNull ServiceRegistration<?>>'. Type 'Map<K,V>' doesn't seem to be designed with null type annotations in mind`

Agree that it not needed. removed it.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/A5_12/A5_12.java [hunk 6]
```diff
@@ -133,6 +135,7 @@ protected State convertToStateImpl(String channelId, String channelTypeId,
                 return getCurrentValue();
             case CHANNEL_TOTALUSAGE:
                 State value = getCumulativeValue();
+                @Nullable
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/A5_12/A5_12.java:138]: Removed several similar cases

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/A5_38/A5_38_08_Dimming.java [hunk 7]
```diff
@@ -55,45 +58,48 @@ public A5_38_08_Dimming(ERP1Message packet) {
...
 
-                byte rampingTime = Integer.valueOf(c.rampingTime).byteValue();
-                byte switchingCommand = (dimmValue == ZERO) ? SwitchOff : SwitchOn;
+                    byte rampingTime = Integer.valueOf(c.rampingTime).byteValue();
+                    byte switchingCommand = (dimmValue == ZERO) ? SWITCH_OFF : SWITCH_ON;
 
-                setData(CommandId, dimmValue, rampingTime, (byte) (TeachInBit | storeByte | switchingCommand));
+                    setData(COMMAND_ID, dimmValue, rampingTime, (byte) (TEACHIN_BIT | storeByte | switchingCommand));
+                } else {
+                    logger.debug("Cannot handle command {}, when configuration is null", outputCommand.toFullString());
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/A5_38/A5_38_08_Dimming.java:101]: Changed it to `error`

**Comment by lsiepel** [comment]
> > Warning: Redundant null check: The variable existingKey cannot be null at this location.
> > I wonder why, because the `putIfAbsent` returns null if the key is not in the receivingEEPTypes.
> 
> I agree that the return value of `putIfAbsent` can be null. @J-N-K is active at lastNPE. Maybe he has an idea, if this is a bug?

Another one: 
```
// EEPType java.util.Hashtable.get(Object key)
EEPType localReceivingType = receivingEEPTypes.get(msg.getRORG());
if (localReceivingType == null) {
// compiler marks this block as dead code.
}
```
The docs say `Hashtable.get` may return `null`
`More formally, if this map contains a mapping from a key k to a value v such that (key.equals(k)), then this method returns v; otherwise it returns null. (There can be at most one such mapping.)`

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/Generic/GenericEEP.java [hunk 8]
```diff
@@ -81,64 +83,68 @@ public GenericEEP(ERP1Message packet) {
...
+
+        String input = channelId + "|" + command.toString();
+        String output = Transformation.transform(transformationInfo.transformationType,
+                transformationInfo.transformationFunction, input);
+
+        if (output != null && !output.isEmpty() && !input.equals(output)) {
+            try {
+                setData(HexUtils.hexToBytes(output));
+            } catch (Exception e) {
+                logger.debug("Command {} could not transformed", command.toString());
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/Generic/GenericEEP.java:101]: Fixed several occurences: log a debug message and re-throw the exception.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/handler/EnOceanBaseSensorHandler.java [hunk 9]
```diff
@@ -82,15 +86,20 @@ boolean validateConfig() {
         try {
             config.receivingEEPId.forEach(receivingEEP -> {
                 EEPType receivingEEPType = EEPType.getType(receivingEEP);
-                if (receivingEEPTypes.putIfAbsent(receivingEEPType.getRORG(), receivingEEPType) != null) {
+                @Nullable
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/handler/EnOceanBaseSensorHandler.java:89]: Are you sure? putIfAbsent can return null. If i remove the `@Nullable` the compiler outputs:
`Redundant null check: The variable existingKey cannot be null at this location`
But that is not true. My knowledge to these warnings and the compiler actually making code path's is limited. Why i don’t want is an NPE at runtime because the compiler left this code path out of the equation.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/messages/Responses/RDLearnedClientsResponse.java [hunk 10]
```diff
@@ -39,10 +41,10 @@ public RDLearnedClientsResponse(Response response) {
     RDLearnedClientsResponse(int dataLength, int optionalDataLength, byte[] payload) {
         super(dataLength, optionalDataLength, payload);
 
-        if (payload == null || ((payload.length - 1) % 9) != 0) {
+        if (payload.length == 0 && (payload.length - 1) % 9 != 0) {
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/messages/Responses/RDLearnedClientsResponse.java:44]: No, good catch!

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/transceiver/EnOceanESP2Transceiver.java [hunk 11]
```diff
@@ -128,30 +137,35 @@ protected void processMessage(byte firstByte) {
...
+                            dataBuffer[currentPosition++] = byteBuffer;
                         }
                         break;
                 }
             }
         } catch (
 
         IOException ioexception) {
-            errorListener.ErrorOccured(ioexception);
+            logger.trace("Unable to process message", ioexception);
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/transceiver/EnOceanESP2Transceiver.java:164]: Because the `localListener.errorOccured(ioexception);` sets the thing offline with an error description for the end user. I did not follow all possible call hierarchies to check if any IOException is throw, they might be there, if so this log call would help troubleshoot.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/transceiver/EnOceanESP3Transceiver.java [hunk 12]
```diff
@@ -175,13 +184,17 @@ protected void processMessage(byte firstByte) {
...
                         } else {
-                            dataBuffer[currentPosition++] = _byte;
+                            dataBuffer[currentPosition++] = byteBuffer;
                         }
                         break;
                 }
             }
         } catch (IOException ioexception) {
-            errorListener.ErrorOccured(ioexception);
+            logger.trace("Unable to process message", ioexception);
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/transceiver/EnOceanESP3Transceiver.java:193]: See other comment. Because the messages are similar i think a stacktrace is added value.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/transceiver/EnOceanTransceiver.java [hunk 13]
```diff
@@ -101,32 +106,42 @@ private synchronized void send() throws IOException {
...
                                 }
-                            }, 250, TimeUnit.MILLISECONDS);
+
+                                // slowdown sending of message to avoid hickups at receivers
+                                // Todo tweak sending intervall (250 ist just a first try)
+                                timeOut = scheduler.schedule(() -> {
+                                    try {
+                                        sendNext();
+                                    } catch (IOException e) {
+                                        logger.trace("Unable to process message", e);
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/transceiver/EnOceanTransceiver.java:136]: See other comment. Because the messages are similar i think a stacktrace is added value.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/transceiver/EnOceanTransceiver.java [hunk 14]
```diff
@@ -201,19 +235,26 @@ public void run() {
...
+        Future<?> localReadingTask = readingTask;
+        if (localReadingTask != null) {
+            localReadingTask.cancel(true);
+
+            InputStream localInputStream = inputStream;
+            if (localInputStream != null) {
+                try {
+                    localInputStream.close();
+                } catch (IOException e) {
+                    logger.debug("IOException occured while closing the stream", e);
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/transceiver/EnOceanTransceiver.java:256]: Yes, because this is such a large binding and this message could be duplicated elsewhere making it hard to troubleshoot. Putting the binding debug logging is usually per developer request. So i don't expect end-users logs to be bloated.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/transceiver/EnOceanTransceiver.java [hunk 15]
```diff
@@ -224,26 +265,28 @@ public void ShutDown() {
...
-        if (outputStream != null) {
-            logger.debug("Closing serial output stream");
+        OutputStream localOutputStream = outputStream;
+        if (localOutputStream != null) {
             try {
-                outputStream.close();
+                localOutputStream.close();
             } catch (IOException e) {
-                logger.debug("Error while closing the output stream: {}", e.getMessage());
+                logger.debug("IOException occured while closing the output stream", e);
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/transceiver/EnOceanTransceiver.java:273]: Check similar comment

**Comment by lsiepel** [comment]
> The folder of the generic package is still upper case.

yeah, great! windows/ vsccode/ sat all reported `generic`, anyway, fiddled around and got it fixed.

Great thanks for the review, have learned alot again. Some comments are still open. I also update the JAR as i think some additionall testing is needed. On the other hand, with all the breaking changed and java17, i don't think anyway will actively be ttesting this binding in 4.0.0.

**Review by fwolter — APPROVED** [review]
LGTM

Can you resolve the merge conflicts?

Please note that we don't do functional tests during review. It's obliged to you to test the changes thoroughly or find someone in the community to help testing. Due to the amount of changed lines in this PR, regressions could have been easily introduced. Are you sure you tested the changes sufficiently?

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/A5_3F/A5_3F_7F_Universal.java [hunk 16]
```diff
@@ -41,21 +44,21 @@ public A5_3F_7F_Universal(ERP1Message packet) {
...
-
-                if (c != null && !c.equals(command.toString())) {
-                    setData(HexUtils.hexToBytes(c));
-                }
-
-            } catch (Exception e) {
-                logger.debug("Command {} could not transformed", command.toString());
+            Function<String, State> getCurrentStateFunc, @Nullable Configuration config) {
+        if (config == null) {
+            logger.debug("Command {} could not transformed without proper configuration", command.toString());
```
[the remark this answers is not in the document]
Comment (lsiepel) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/A5_3F/A5_3F_7F_Universal.java:49]: Missed this, changed it to error.

**Comment by lsiepel** [comment]
> LGTM
> 
> Can you resolve the merge conflicts?
> 
> Please note that we don't do functional tests during review. It's obliged to you to test the changes thoroughly or find someone in the community to help testing. Due to the amount of changed lines in this PR, regressions could have been easily introduced. Are you sure you tested the changes sufficiently?

Think you are right and it would be better to do some additional testing. I'll hunt some people in the community and will come back if it is ready to merge. In the mean while i'll try to get these conflicts fixed. Usually they end up in disaster :-(

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/A5_3F/A5_3F_7F_Universal.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (fwolter) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/A5_3F/A5_3F_7F_Universal.java:49]: It's still debug.

**Review by fwolter — CHANGES_REQUESTED** [review]

**Comment by kaikreuzer** [comment]
@fwolter Friendly ping - as you had almost finished this review, could you check the final commits from @lsiepel that addressed your review comments?
